Tools/Scripts/layout-test-failure-analysis

```diff
@@ -0,0 +1,291 @@
+                        dest="stdio_export_path",
+                        help='Specifies a path to which to save the raw stdio received from the server. This has no effect when loading data from a file.',
+                        required=False)
+    parser.add_argument('--combine-workers',
+                        action='store_true',
+                        help='Exports one file containing all tests in the order they were run (regardless of worker number).',
+                        required=False)
+    parser.add_argument('--no-warn',
+                        action='store_true',
+                        help='Does not warn or prompt the user before performing any actions. Potentially destructive due to bypassing file overwrite prompts. Good for automation. Will cause the script to quit if user input required.',
+                        required=False)
+
+    return parser.parse_args()
+
+
+# logging shortcuts
+def info(msg):
+    logging.info(msg)
+def warning(msg):
+    logging.warning(f'WARNING: {msg}')
+def error(msg):
+    logging.error(f'ERROR: {msg}')
+
+
+def graceful_exit(code=0):
+    print()
+    sys.exit(code)
+
+
+# Ensures a directory exists by checking and attempting to create if it doesn't exist (recursively).
+# Returns true if it exists at the end of the attempt, false if not.
+def ensure_directory_exists(dir):
+    if not os.path.isdir(dir):
+        os.makedirs(dir, exist_ok=True)
+        if not os.path.isdir(dir):
+            return False
+    return True
+
+
+# Takes a string and exports to specified file path.
+# Prompts the user for a new file path if there's an IO error.
+def export_file(path, data, no_warn = False):
+    path
+    while True:
+        try:
+            path = os.path.abspath(os.path.expanduser(path))
+            dir = os.path.dirname(path)
+            if os.path.isfile(path) and not no_warn: # file exists
+                choice = input(f'A file already exists at the location {path}. Is it okay to overwrite it? (Y/[N]): ').lower()
+                if choice.lower() != 'y' and choice.lower() != 'yes':
+                    raise ValueError()
+            if not ensure_directory_exists(dir): # directory doesn't exist
+                warning(f'Directory does not exist at {dir} and an error occurred while attempting to create it. Check permissions/spelling and try again.')
+                raise IOError()
+
+            # write file
+            print(path)
+            f = open(path, "w")
+            f.write(data)
+            f.close()
+            return path
+        except IsADirectoryError:
+            warning(f'The specified path "{path}" is a directory. Try entering a new file path.')
+        except OSError:
+            warning(f'The specified location "{path}" is read-only. Try entering a new file path.')
+        except ValueError as e:
+            pass
+        except Exception as e:
+            warning(f'Unable to write file: {e.args[1]}. Try entering a new file path.')
```

Comment:
Again, no reason to use `as e` if we aren't going to use the `e`. We usually avoid catching generic `Exception`s, unless we're working in code which is managing other code (like our multiprocessing code, for example). The error message kind of demonstrates why. We really have no way of knowing what the raised exception is about, so we can't really suggest recovery steps.

Reply:
`e` is used in the string, `f'Unable to write file: {e.args[1]}. Try entering a new file path.'`.

The idea here is that, if the input file has already been downloaded and parsed, and all that needs to be done is write the output to a file, we don't want to crash and make the user start all over again. After the exception, the user is prompted to enter a new file path.